### DISCUSSION TITLE
Merge the "withdraw" with the "withdraw to" functions

### DISCUSSION THREAD

**Discussion by PaulRBerg** [sablier-labs/evm-monorepo#192] [Ideas]
There's quite a bit of repetitive code in the current withdraw functions, which is mainly due to the plurality of functions we have.

What if we merged the basic "withdraw" and the "withdraw to" variants, like this?

- We would only have the `withdraw` and `withdrawAll` functions, but each would have an additional argument `to`
- `to` can be any other address, but only if the caller is the recipient or an approved operator thereof
- If the caller is the sender but `to` is not the recipient, the function reverts

This design would let us delete two functions and two large test suites, with the only cost being the slight added cost for checking the new input argument requirements.

**Comment by andreivladbrg** [comment]
I like the idea. The code reduction would be nice.

**Reply by PaulRBerg** [reply]
I'm glad that you like it. Purging these functions might actually be necessary - as I have just seen in this [CI report](https://github.com/sablierhq/v2-core/actions/runs/3669687625/jobs/6203691460), the `forge build --sizes` command failed in PR #193 due to the `SablierV2Pro` contract having gone above the 24kB size limit.

**Comment by andreivladbrg** [comment]
Is the `to` address [check](https://github.com/sablierhq/v2-core/blob/2445865e3269481fb47fb320e71d2e1f42382cf6/src/SablierV2.sol#L217-L219) still necessary in the `withdraw` function? The ERC20 transfer function would revert with [this](https://github.com/paulrberg/prb-contracts/blob/321a346bfcbc6d77f85a0bdab4348f06a206d7ff/src/token/erc20/ERC20.sol#L221-L223) reason.

**Reply by PaulRBerg** [reply]
Good catch. I don't think we need to keep the check if it is also implemented in the ERC-20 contract itself. However, I would like us to keep the test (and update the custom errors, accordingly), because we should definitely forbid withdrawals to the zero address.

**Comment by PaulRBerg** [comment]
Locking, implemented in #205.
